cocotb/bfmgen.py

```diff
@@ -0,0 +1,398 @@
+    out_c.write("//* Note: This file is generated. Do Not Edit\n")
+    out_c.write("//***************************************************************************\n")
+    out_c.write("#include <stdio.h>\n")
+    out_c.write("#ifdef __cplusplus\n")
+    out_c.write("extern \"C\" {\n")
+    out_c.write("#endif\n")
+    out_c.write("\n")
+    out_c.write("#include \"svdpi.h\"\n")
+    out_c.write("typedef void (*cocotb_bfm_notify_f)(void *);\n")
+    out_c.write("int cocotb_bfm_register(const char *, const char *, const char *, cocotb_bfm_notify_f, void *);\n")
+
+    for t in inst.bfm_type_info_m.keys():
+        info = inst.bfm_type_info_m[t]
+        
+        if cocotb.bfm_vlog not in info.hdl.keys():
+            raise Exception("BFM \"" + t.__name__ + "\" does not support Verilog")
+        
+        template_f = open(info.hdl[cocotb.bfm_sv], "r")
+        template = template_f.read()
+        template_f.close()
+        
+        bfm_name = os.path.basename(info.hdl[cocotb.bfm_sv])
+    
+        if bfm_name.find('.') != -1:
+            bfm_name = os.path.splitext(bfm_name)[0]
+        
+        out_sv.write(process_template_sv(template, bfm_name, info))
+        out_c.write(generate_dpi_c(bfm_name, info))
+        
+    out_c.write("#ifdef __cplusplus\n")
+    out_c.write("}\n")
+    out_c.write("#endif\n")
+    
+    out_sv.close() 
+    out_c.close()
+
+def bfm_generate(args):
+    '''
+    Generates BFM files required for simulation
+    '''
+    
+    if args.o is None:
+        if args.language == "vlog":
+            args.o = "cocotb_bfms.v"
+        elif args.language == "sv":
+            args.o = "cocotb_bfms.sv"
+        elif args.language == "vhdl":
+            args.o = "cocotb_bfms.vhdl"
+            
+    if args.language == "vlog":
+        bfm_generate_vl(args)
+    elif args.language == "sv":
+        bfm_generate_sv(args)
+    elif args.language == "vhdl":
+        args.o = "cocotb_bfms.vhdl"
+            
+    
+def bfm_list(args):
+    print("bfm_list")
+    inst = BfmMgr.inst()
+    
+    print("inst=" + str(inst))
+
+    print("Number of keys: " + str(len(inst.bfm_type_info_m.keys())))
+    for t in inst.bfm_type_info_m.keys():
+        print("BFM: \"" + str(t) + "\"")
+
+def main():
+    parser = argparse.ArgumentParser(prog="cocotb-bfms")
+    
+    subparser = parser.add_subparsers()
+    generate_cmd = subparser.add_parser("generate")
+    generate_cmd.set_defaults(func=bfm_generate)
+    generate_cmd.add_argument("-m", action='append')
+    generate_cmd.add_argument("-language", default="vlog")
+    generate_cmd.add_argument("-o", default=None)
+    
+    list_cmd = subparser.add_parser("list")
+    list_cmd.set_defaults(func=bfm_list)
+    list_cmd.add_argument("-m", action='append')
+    
+    args = parser.parse_args()
+ 
+    # Ensure the BfmMgr is created
+    BfmMgr.inst()
+    
+    if hasattr(args, 'm') and args.m is not None:
+        bfm_load_modules(args.m)
+        
+    args.func(args)
```

Comment:
If I call ``cocotb-bfmgen`` without any arguments or options, Python says about this line:
``AttributeError: 'Namespace' object has no attribute 'func'``

Reply:
Good catch! Just pushed a fix